src/v/iceberg/rest_client/retry_policy.h

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "http/client.h"
+#include "iceberg/rest_client/types.h"
+
+namespace iceberg::rest_client {
+
+// Repreents an http call which failed, encodes information about possible
+// retriability and cause
+struct failure {
+    bool can_be_retried;
+    http_call_error err;
+};
+
+struct retry_policy {
+    using result_t = tl::expected<http::downloaded_response, failure>;
+
+    // Given a ready future which will yield a downloaded_response, judges
+    // whether it is:
+    // 1. successful
+    // 2. has failed and can be retried
+    // 3. has failed and cannot be retried
+    virtual result_t
+    should_retry(ss::future<http::downloaded_response> response_f) const
+      = 0;
+
+    // If the ready future did not fail, this method checks the response http
+    // status
+    virtual result_t should_retry(http::downloaded_response response) const = 0;
+
+    // Handles the case where the future failed with an exception
+    virtual failure should_retry(std::exception_ptr ex) const = 0;
+    virtual ~retry_policy() = default;
+};
+
+struct default_retry_policy : public retry_policy {
```

Comment:
Is the idea that the caller will retry forever? What about backoff?

Reply:
The idea is that the retry policy only controls whether we should retry based on status and exceptions. 

An additional input to the decision is a retry chain node which the caller is expected to consult before making a retry attempt, for a concrete example see https://github.com/redpanda-data/redpanda/blob/2629d134becf30fdced1be5f284b202418dc060a/src/v/iceberg/rest_client/catalog_client.cc#L137-L140

Initially I thought of adding the retry chain node as an input to the retry policy, and checking its permit within should_retry, but to me it seemed cleaner to just let the policy examine the status code and possible exceptions and let the caller check its retry chain node. 

It should still be possible to pass the rtc to the policy though, if that seems more readable.

Reply:
I just realized that I had missed adding a backoff in the catalog client. Added now (for retriable errors)